src/AutoPilotPlugins/APM/APMSensorsComponent.qml

```diff
@@ -411,11 +412,33 @@ SetupPage {
                 QGCViewDialog {
                     id: orientationsDialog
 
+                    function compassMask () {
+                        var mask = 0
+                        mask |=  (0 + (sensorParams.rgCompassPrio[0].rawValue !== 0)) << 0
+                        mask |=  (0 + (sensorParams.rgCompassPrio[1].rawValue !== 0)) << 1
+                        mask |=  (0 + (sensorParams.rgCompassPrio[2].rawValue !== 0)) << 2
+                        return mask
+                    }
+
                     function accept() {
                         if (_orientationDialogCalType == _calTypeAccel) {
                             controller.calibrateAccel(_doSimpleAccelCal)
                         } else if (_orientationDialogCalType == _calTypeCompass) {
-                            controller.calibrateCompass()
+                            if (!northCalibrationCheckBox.checked) {
+                                controller.calibrateCompass()
+                            } else {
+                               var lat = parseFloat(northCalLat.text)
+                               var lon = parseFloat(northCalLon.text)
+                               if (useMapPositionCheckbox.checked) {
+                                   lat = _mapPosition.latitude
+                                   lon = _mapPosition.longitude
+                               }
+                               if (useGcsPositionCheckbox.checked) {
+                                   lat = _gcsPosition.latitude
+                                   lon = _gcsPosition.longitude
+                               }
+                               controller.calibrateCompassNorth(lat, lon, compassMask())
```

Comment:
You should check if lat and lon are `NaN`, since `parseFloat` (and maybe mapPosition or gcsPotion) could return invalid values.